sway/security.c

```diff
@@ -1,13 +1,50 @@
 #define _XOPEN_SOURCE 500
+#include <sys/types.h>
+#include <sys/stat.h>
 #include <string.h>
 #include <unistd.h>
 #include <stdio.h>
 #include "sway/config.h"
 #include "sway/security.h"
+#include "util.h"
 #include "log.h"
 
+static bool validate_ipc_program_target(const char *program) {
+	struct stat sb;
+
+	if (!strcmp(program, "*")) {
+		return true;
+	}
+	if (lstat(program, &sb) == -1) {
+		goto failed;
+	}
+	if (!S_ISREG(sb.st_mode)) {
+		sway_log(L_ERROR,
+				"IPC target '%s' MUST be/point at an existing regular file",
+				program);
+		goto failed;
+	}
+	if (sb.st_uid != 0) {
+		sway_log(L_ERROR, "IPC target '%s' MUST be owned by root", program);
+		goto failed;
+	}
+	if (sb.st_mode & S_IWOTH) {
+		sway_log(L_ERROR, "IPC target '%s' MUST NOT be world writable", program);
+		goto failed;
+	}
+
+	return true;
+
+failed:
+	return false;
```

Comment:
No need for the failed label and all the gotos in this function, just return false wherever you would goto failed